Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+        if (_module_state != NULL) {
+            set_zstd_error(_module_state, ERR_FINALIZE_DICT, zstd_ret);
+        }
+        goto error;
+    }
+
+    /* Resize dict_buffer */
+    if (_PyBytes_Resize(&dst_dict_bytes, zstd_ret) < 0) {
+        goto error;
+    }
+
+    goto success;
+
+error:
+    Py_CLEAR(dst_dict_bytes);
+
+success:
+    PyMem_Free(chunk_sizes);
+    return dst_dict_bytes;
+}
+
+
+/*[clinic input]
+_zstd._get_param_bounds
+
+    is_compress: bool
+        True for CParameter, False for DParameter.
+    parameter: int
+        The parameter to get bounds.
+
+Internal function, get CParameter/DParameter bounds.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd__get_param_bounds_impl(PyObject *module, int is_compress,
+                             int parameter)
+/*[clinic end generated code: output=b751dc710f89ef55 input=fb21ff96aff65df1]*/
+{
+    ZSTD_bounds bound;
+    if (is_compress) {
+        bound = ZSTD_cParam_getBounds(parameter);
+        if (ZSTD_isError(bound.error)) {
+            _zstd_state* const _module_state = PyModule_GetState(module);
+            if (_module_state != NULL) {
+                set_zstd_error(_module_state, ERR_GET_C_BOUNDS, bound.error);
+            }
+            return NULL;
+        }
+    } else {
+        bound = ZSTD_dParam_getBounds(parameter);
+        if (ZSTD_isError(bound.error)) {
+            _zstd_state* const _module_state = PyModule_GetState(module);
+            if (_module_state != NULL) {
+                set_zstd_error(_module_state, ERR_GET_D_BOUNDS, bound.error);
+            }
+            return NULL;
+        }
+    }
+
+    return Py_BuildValue("ii", bound.lowerBound, bound.upperBound);
+}
+
+/*[clinic input]
+_zstd.get_frame_size
+
+    frame_buffer: Py_buffer
+        A bytes-like object, it should start from the beginning of a frame,
+        and contains at least one complete frame.
+
+Get the size of a zstd frame, including frame header and 4-byte checksum if it has one.
+
+It will iterate all blocks' headers within a frame, to accumulate the frame size.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd_get_frame_size_impl(PyObject *module, Py_buffer *frame_buffer)
+/*[clinic end generated code: output=a7384c2f8780f442 input=7d3ad24311893bf3]*/
+{
+    size_t frame_size;
+    PyObject *ret;
+
+    frame_size = ZSTD_findFrameCompressedSize(frame_buffer->buf, frame_buffer->len);
+    if (ZSTD_isError(frame_size)) {
+        _zstd_state* const _module_state = PyModule_GetState(module);
+        if (_module_state != NULL) {
+            PyErr_Format(_module_state->ZstdError,
+                "Error when finding the compressed size of a zstd frame. "
+                "Make sure the frame_buffer argument starts from the "
+                "beginning of a frame, and its length not less than this "
+                "complete frame. Zstd error message: %s.",
+                ZSTD_getErrorName(frame_size));
+        }
+        goto error;
```

Comment:
Just return NULL here and then `return PyLong_FromSize_t(frame_size);`:

```c
frame_size = ....
if (ZSTD_isError(...)) {
   ...
   return NULL;
}
return PyLong_FromSize_t(frame_size);
```

Reply:
Much neater, thanks!